controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,856 @@
+/* Copyright 2020, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#if defined(BARE_STM32)
+
+#include "hal.h"
+#include "stm32.h"
+
+// This is the main stack 
+uint32_t mainStack[ 500 ];
+
+// local data 
+static volatile uint32_t msCount;
+
+// local static functions.  I don't want to add any private 
+// functions to the Hal class to avoid complexity with other
+// builds
+static void InitGPIO( void );
+static void InitADC( void );
+static void InitSysTimer( void );
+static void InitPwmOut( void );
+static void InitUARTs( void );
+static void EnableClock( uint32_t base );
+static void EnableInterrupt( int addr, int pri );
+static void DisableInterrupt( int addr );
+static void Timer6ISR( void );
+static void UART3_ISR( void );
+
+// For now, the main function in main.cpp is called setup
+// rather then main.  If we adopt this HAL then we can 
+// just rename it main and get rid of the following function.
+extern void setup( void );
+int main( void ){
+   setup();
+}
+
+
+// This function is called from the libc initialization code 
+// before any static constructors are called.  We do some basic
+// chip initialization here.
```

Comment:
I added some additional comments to the code.
Basically, this init function is called before the static constructors are called.  There's some basic chip init that needs to be done here to prevent faults.  I also initialize the clock so the constructors run at 80MHz rather then 4MHz which is the default clock on power-up.
I don't initialize all the chips peripherals here because I don't expect any static constructors to be using the HAL, the HAL in general hasn't been initialized yet.  
Do you guys think we should make that a rule (don't access the HAL in static constructors) or do you think that all the basic HAL init should be done at this point?